Refactor/straight  build functions

progfolio opened the pull request:
An implementation of what I was suggesting with #659 and #655.
The README will need to be reworded, but I figured we could discuss the viability of such an approach before making substantial edits.

A recipe may specify build steps via the `:build` keyword.
(Note the original `:build` keyword has been renamed to `:pre-build`)

The `:build` keyword accepts the following values:

- `nil`
No build is executed. `:pre-build` and `:post-build` commands are not run.

```emacs-lisp
     (example :build nil)
```

- `t`
Equivalent to running `straight--default-build-steps` (see below).

- `(step...)`
Each step in the list is a symbol which represents a function named `straight--build-SYMBOL`.
The function is passed the recipe as its sole argument.
Steps are executed in the order they are listed.

```emacs-lisp
(example :build (autoloads compile native-compile info))
```

- `(:not step...)`
  If the car of the list is the `:not` keyword, the build steps are the
  set difference of `straight--build-default-steps` and the cdr of the
  recipe's `:build` list.
  e.g. with `straight--build-default-steps` set to:

```emacs-lisp
(autoloads compile native-compile info)
```

   and the following recipe:

```emacs-lisp
(example :build (:not compile info))
```

  the package is not byte compiled and does not have it's texinfo generated.

  Steps may be disabled globally via the customization variable named
  `straight-disable-SYMBOL`. e.g.:

```emacs-lisp
(setq straight-disable-info t)
;;info generation disabled when :build is not explicitly declared
(straight-use-package '(example))
;;or starts with :not
;;Info disabled in addition to compilation.
(straight-use-package '(example :build (:not compile)))
```

In the absence of a `:build` keyword, `straight--build-default-steps` are run.

raxod502 commented:
Yeah, this all sounds perfect! It seems like a much better interface all around. Same with your comments in https://github.com/raxod502/straight.el/pull/655, https://github.com/raxod502/straight.el/issues/654, https://github.com/raxod502/straight.el/issues/659: this all seems great.

progfolio commented:
Sounds good. I'll add some tests, `define-obsolet-variable-alias`es for the option name changes, and edit the documentation. How do you prefer to deal with the news of the `:build` -> `:pre-build` rename? Is a notice in the README sufficient?

progfolio commented:
Added a unit test for `straight--build-setps`, properly obsoleted changed defcustoms, updated README and made note of the breaking change in the NEWS section.

raxod502 commented:
> Is a News entry in the README sufficient?

I think that would be fine, given that `:build` itself was only added recently.
